### ISSUE TITLE
Question about forwarding request information to the adapter

### ISSUE THREAD

**Issue by dalefukami** [fortunejs/fortune#92]
I'm looking for a way to push the request cookies down to the adapter level and I'm wondering if someone might have some suggestions.

We're using fortunejs as a sort of middle-man api between a front end component and some existing back end services. So, we've created a custom adapter to access these existing backend services. However, there's some information in the cookies that we need to pass along.

My first glance leads me to believe there are two stages to this.
1. Add the cookies to the meta field somehow
2. Pass the meta down to the various dispatched methods on the adapter. Perhaps as the last parameter after 'options' or as a field on the options hash.  For example: `if (meta) args.push(meta) )`

We'd be happy to submit a pull requests but would like some suggestions prior to getting too deep. Does this sound like something that can be down outside fortunejs in the various plugins or would actual framework changes be required?

Thanks!

**Comment by 0x8890** [comment]
HTTP headers are already appended to the `meta` field if you are using the built-in http handler: https://github.com/fortunejs/fortune/blob/master/lib/net/http.js#L67

Then in an transform function you can access headers like so:

```js
store.transformInput('type', (context, record) => {
  const cookie = context.request.meta['cookie']
  // do something with the cookie
})
```

I don't think any changes need to be made to fortune for working with headers, but I'd be interested in knowing more about your use case.

**Comment by dalefukami** [comment]
Ah, yes, I do see the cookie within the `meta` field.

In regards to the `transformInput`, I'd like to be able to have this accessible to every method in the adapter. For example, our `find` is reaching out to a web service and we'd need to do something along these lines:
```
http.request({headers: { 'Cookie': <the original cookie> }}, ...)
```

So the issue is in accessing the cookie from within the adapter. From what I can see, all we receive in our adapter for the `find` method are the `type`, `id`, and `options`. Could I add the cookie to the `options` hash within the transformInput? Does the transformInput even actually run for GET requests?

**Comment by 0x8890** [comment]
Sorry I don't understand your use case, so I will just give some general tips. The `transformInput` function is run whenever *writing* a record (all methods except finding, in other words GET for http), and `transformOutput` is run on every request. If the Cookie header doesn't modify anything that needs to be persisted, then you can just do everything in `transformOutput`. However, `transformOutput` is run *after* the record has already been found, so that may not be applicable in your use case.

It is fine to pass any arbitrary field in any `options` object, what you can do is extend the `processRequest` method in the serializer:

```js
processRequest () {
  const [ context, request ] = arguments
  context.request.options['My special option'] = request.headers['cookie']
  return super(...arguments)
}
```

It's actually the serializer that may be coupled with HTTP, and it's what allows completely different behavior per media type.

**Comment by dalefukami** [comment]
Yep, that looks exactly like what I want. I was just toying with extending processRequest :)

Sorry to be a pain, but I'm not quite understanding the docs well enough to figure out how to just extend the JSONAPI serializer. Is there an example kicking around where one of the default serializers is extended and how to use that in place of the original?

**Comment by 0x8890** [comment]
here's an example:

```js
// this works for now, need to expose base class though
import Serializer from 'fortune/dist/lib/adapter'
import fortune from 'fortune'

const JsonApiSerializer = fortune.serializers.JSONAPI(Serializer)
class MySerializer extends JsonApiSerializer {
  process request () { ... }
}

const store = fortune.create({
  serializers: [ { type: MySerializer } ]
})
```

**Comment by dalefukami** [comment]
Hmm, yeah, I had been trying that. I'm getting this:
```
..../node_modules/fortune/dist/lib/serializer/singleton.js:79
f (Object.getPrototypeOf(CustomSerializer) !== _2['default']) throw new TypeEr
                                                                    ^
TypeError: The serializer must be a class that extends Serializer.
```

I presume this is because it's not a direct descendant of Serializer but I haven't googled the details yet.

**Comment by 0x8890** [comment]
Oops, that's a bug then, it should traverse the prototype chain. I will push a fix :)

**Comment by 0x8890** [comment]
Fixed in commit f0fcaae44e8385570c525dc4cd1636bc7ac47952, published as `1.0.0-beta.28`.

**Comment by dalefukami** [comment]
Awesome! Looks like the mechanism of extending the serializer is going to work for us. Thanks so much for the assistance.
